### ISSUE TITLE
SerializationCoderFromType does not work with polymorphic types

### ISSUE THREAD

**Issue by rdubya** [jrochkind/attr_json#205]
Hello!

Thank you for this gem. I am probably trying to use it in a way that isn't yet supported, so maybe the answer to this is that I need to change how I'm using it, but I wanted to get this on your radar if it is something that should work.

The core issue appears to be that when using the SerializationCoderFromType it doesn't handle store keys correctly. Here is a test case:

```
require 'attr_json'

class Test
  include AttrJson::Model

  attr_json :name, :string, store_key: 'n'
end

polymorphic_type = AttrJson::Type::PolymorphicModel.new(Test)

test_coder = Test.to_serialization_coder
polymorphic_coder = AttrJson::SerializationCoderFromType.new(polymorphic_type)

puts 'Test coder'
coder_output = test_coder.dump(Test.new(name: 'test'))
puts coder_output.inspect
result = test_coder.load(coder_output) # => successfully re-hydrates the model
puts result.inspect

puts 'Polymorphic coder'
polymorphic_coder_output = polymorphic_coder.dump(Test.new(name: 'test'))
puts polymorphic_coder_output.inspect
result = polymorphic_coder.load(polymorphic_coder_output) # => fails
puts result.inspect
```

Result:
```
Test coder
{"n"=>"test"}
#<Test:0x0000aaaaf9167430 @attributes={"name"=>"test"}>
Polymorphic coder
{"n"=>"test", "type"=>"Test"}
/bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/model.rb:452:in `_attr_json_write_unknown_attribute': unknown attribute 'n' for Test. (ActiveModel::UnknownAttributeError)
	from /bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/model.rb:314:in `block in assign_attributes'
	from /bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/model.rb:309:in `each'
	from /bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/model.rb:309:in `assign_attributes'
	from /bundle/ruby/3.0.0/gems/activemodel-6.1.7.3/lib/active_model/model.rb:81:in `initialize'
	from /bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/model.rb:285:in `initialize'
	from /bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/type/model.rb:44:in `new'
	from /bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/type/model.rb:44:in `cast'
	from /bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/type/polymorphic_model.rb:174:in `cast_from_hash'
	from /bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/type/polymorphic_model.rb:108:in `cast'
	from /bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/type/polymorphic_model.rb:131:in `deserialize'
	from /bundle/ruby/3.0.0/gems/attr_json-2.0.1/lib/attr_json/serialization_coder_from_type.rb:37:in `load'
	from test_attr_json_coder.rb:23:in `<main>'
```

I've tracked it to the point where I can see that is effectively trying to do `Test.new({"n"=>"test"})` when it fails. I'm not sure if it should work to pass in a hash like that or if there is a step in the load process for polymorphic types that needs to be adjusted to fix those keys first.

**Comment by jrochkind** [comment]
Hi, thanks for trying this out and giving a problem report, with a close-to test case. 

It wouldn't surprise me that the `SerializationCoderFromType` feature is a bit not quite baked. I added it fairly late in the game in reply to user request, and don't use it myself. 

* I am not sure if I'm going to have time to look at this soon, but if you wanted to keep experimenting and provide a PR that would fix this for the `SerializationCoderFromType` approach, I would be happy to look at it!

Additionally, though, I realized after the fact that Rails is maybe moving away from "Serialization Coders", the feature I tried to use to implement this, and instead we could have just supplied a "cover" type with `attribute`?  

There was a reason you couldn't just use our existing `AttrJson::Type::Model` type directly with ActiveRecord and a json column, something like this:

    attribute :my_json_column, MyAttrJsonModel.to_type

It involved that our type serialized/deserialized only to a JSON hash, when that approach wants it all the way to a string -- or maybe it was vice/versa I forget. 

* But another appraoch would be trying to provide an alternate ActiveModel::Type for our `Model::Type` that serialized the appropriate way for using directly with AR `attribute` as above. I think that's actually the more modern ActiveRecord way to do this than with the SerializationCoder we are using, and how I would have done it if I had realized this -- so instead of bugfixing the SerializationCoder, perhaps you want to experiment with that approach, and see if it exhibits the same problem, and bug fix that if needed?

**But finally**, seeing that your problem is specifically with polymorphic types too... I wonder if there's going to be an essential problem there, where it gets complicated to properly use the store keys from the approrpriate type on deserialization, when multiple types are possible? Like it needs to FIRST identify the type, THEN make sure it's using the proper store key definitions from that type, and maybe that doesn't fit well into the logic flow? **But that might not be an issue, I haven't looked at the code.**

Basically, I'm not sure when I'm going to have time to dig into this in code myself, but I'm happy to continue to give you advice and feedback and consider PR's if you'd like to keep looking into it?
